### PR TITLE
#730 add user settings

### PR DESCRIPTION
**Pull request by jamescd18**
## Changes

Added user settings to the database model and prisma and seed data.

Okay so basically my plan here was to separate out and have the login endpoint be what first returns the user settings, since this would avoid an extra call to the server and allow us to set up the app with the user's preferred settings as soon as they log in.

This also goes along with my plan to change around the user endpoints to remove sending unnecessary info to the front-end whenever possible. Having a singular unified User interface makes that somewhat difficult, so this take a first step to try and have a separate interface for authenticated user and see how things go.

## Screenshots

Screen Shot 2022-07-02 at 3 42 11 AM

Screen Shot 2022-07-02 at 3 42 28 AM

Screen Shot 2022-07-02 at 3 42 52 AM

instant change post-change
Screen Shot 2022-07-02 at 3 43 17 AM

preserved on reload, theme is loaded on login
Screen Shot 2022-07-02 at 3 43 41 AM

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/docs/ContributorGuide.md) and reach out to your squad if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors
- [x] No newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (if applicable)
- [x] Remove any not-applicable sections
- [x] Assign the PR to yourself
- [x] PR is linked to the ticket
- [x] No `package-lock.json` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack

### CODE REVIEW

File: src/shared/types.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@
  * See the LICENSE file in the repository root folder for details.
  */
 
-import { User } from '@prisma/client';
+import { User } from 'utils';
```
Comment (anthonybernardi) [inline src/shared/types.ts:6]: good catch!

Comment (anthonybernardi) [reply src/shared/types.ts:6]: on all of these didn't realize we did this so much

File: src/backend/prisma/schema.prisma [hunk 2]
```diff
@@ -53,6 +58,9 @@ model User {
   email                   String           @unique
   emailId                 String?          @unique
   role                    Role             @default(GUEST)
+  userSettingsId          String?          @unique
+  userSettings            UserSettings?    @relation(fields: [userSettingsId], references: [id])
```
Comment (anthonybernardi) [inline src/backend/prisma/schema.prisma:62]: I think that the user should have just the userSettings object and the userSettings model should be concerned with the id and @relation tag. I don't know if it matterrs but it makes more rsense to me in the sense that

Comment (anthonybernardi) [inline src/backend/prisma/schema.prisma:62]: I think that the user should have just the userSettings object and the userSettings model should be concerned with the id and @relation tag. I don't know if it matters but it makes more sense to me in the sense that the user settings is then tied to the user

**Review by anthonybernardi — COMMENTED** [review]
i like it. so is the plan then to basically make the authenticated user thing part of a context when logged in so it gets stored and stuff so we only need it once?

File: src/shared/types.ts (hunk 1 above)
Comment (jamescd18) [reply src/shared/types.ts:6]: Lmao me neither!

File: src/backend/prisma/schema.prisma (hunk 2 above)
Comment (jamescd18) [reply src/backend/prisma/schema.prisma:62]: Lemme experiment and I'll see if it's possible, I can't remember what did / didn't work

**Comment by jamescd18** [comment]
Yeah basically. I guess it kinda depends on how we expand user settings and what we store in there. We would separately enable editing of user settings via the settings page.

File: src/backend/prisma/schema.prisma (hunk 2 above)
Comment (jamescd18) [reply src/backend/prisma/schema.prisma:62]: Like this?

Screen Shot 2022-07-02 at 3 40 26 AM

Comment (anthonybernardi) [reply src/backend/prisma/schema.prisma:62]: yeee

File: src/frontend/app/app-public/app-public.tsx [hunk 3]
```diff
@@ -20,7 +20,7 @@ const AppPublic: React.FC = () => {
   document.body.style.backgroundColor = theme.bgColor;
 
   return (
-    <html className={theme.name}>
+    <html className={theme.cardBg}>
```
Comment (anthonybernardi) [inline src/frontend/app/app-public/app-public.tsx:23]: this is not always true in the case of expanded themes -- the name field of the theme is meant to be the class name

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
This is very cool. Here are some initial notes, haven’t gone fully in depth but I’ve looked at everything so far:
- Will want to get rid of the top theme picker
- Thomas emrax’s theme doesn’t load in as light on my local even if it’s set as his default theme
- I think that all users will naturally have no settings because it’s optional and then all the endpoints will not work because they assume it already exists, when they should probably create a default settings if it doesnt exist

File: src/frontend/app/app-public/app-public.tsx (hunk 3 above)
Comment (jamescd18) [reply src/frontend/app/app-public/app-public.tsx:23]: we'll need some kind of translation from a given theme to a unique identifier. what do you suggest?

**Comment by jamescd18** [comment]
> Will want to get rid of the top theme picker

I figured we could keep it. did you think it's better removed? cuz if we remove it then there's technically no way to change your current theme without changing your default theme, right? Is this what you'd want?

> Thomas emrax’s theme doesn’t load in as light on my local even if it’s set as his default theme

hmmm, any ideas why?

> I think that all users will naturally have no settings because it’s optional and then all the endpoints will not work because they assume it already exists, when they should probably create a default settings if it doesnt exist

I was figuring we would fix this by manually generating settings for everyone when we deploy this, which I think is no biggie given the manual script

File: src/frontend/app/app-public/app-public.tsx (hunk 3 above)
Comment (anthonybernardi) [reply src/frontend/app/app-public/app-public.tsx:23]: and then className={theme.className}

**Comment by anthonybernardi** [comment]
ok so I see why it doesn't work on dev signing and im pretttty sure this will work on prod. The one commit I made was just so that it uses the chosen theme instead of the default theme value in the hook

looks good to me!

File: src/backend/functions/users.ts [hunk 4]
```diff
@@ -97,7 +124,38 @@ const logUserIn: ApiRouteFunction = async (_params, event) => {
...
+      where: { userId },
+      data: { userSettings: { create: {} } },
+      include: { userSettings: true }
+    });
+  }
+  return buildSuccessResponse(user.userSettings!);
+};
+
+/** Update settings for the specified user */
+const updateUserSettings: ApiRouteFunction = async (params: { id: string }, event) => {
```
Comment (anthonybernardi) [inline src/backend/functions/users.ts:149]: what if this user doesn't have settings yet? I guess we're relying on the user going to the settings page first which should call "getUserSettings" for them and create the settings object

**Review by anthonybernardi — COMMENTED** [review]
our auth system is kinda elaborate

File: src/backend/functions/users.ts (hunk 4 above)
Comment (jamescd18) [reply src/backend/functions/users.ts:149]: correct. I'm still maybe sorta a fan of just editing the data so there is no chance for a user to ever not have a user settings object. ignoring that though, the get request will create it. and the update form can't be accessed without calling the get request, so it's guaranteed to always exist for the update. the only scenario possible is if someone tries to update via API access. we could change to upsert. we would still need an answer for what to return if get has none though. lmao the more you know. halfway through writing this I googled it and: https://github.com/prisma/docs/issues/640

so basically either we
- edit data and assume it's always there
- leave it as is and say meh
- change both to upsert cuz that's most robust

Comment (jamescd18) [reply src/backend/functions/users.ts:143]: > change both to upsert cuz that's most robust

I picked this one cuz it's more robust and actually less code I think

File: src/services/theme.hooks.ts [hunk 5]
```diff
@@ -6,10 +6,13 @@
...
 import { ThemeContext } from '../frontend/app/app-context-theme/app-context-theme';
 import themes from '../shared/themes';
+import { useAuth } from './auth.hooks';
 
 // Provider hook that creates theme object and handles state
 export const useProvideTheme = () => {
-  const [theme, setTheme] = useState(themes[0]);
+  const auth = useAuth();
+  const defaultThemeIndex = themes.findIndex((t) => auth.user?.defaultTheme === t.name);
+  const [theme, setTheme] = useState(themes[defaultThemeIndex > -1 ? defaultThemeIndex : 0]);
```
Comment (jamescd18) [inline src/services/theme.hooks.ts:15]: Sorta still confused how this helps / what this intended to do? Is this intended to solve the dev user default theme not working?

The way I set it up was
1. visit login page
2. default dark always used via index 0
3. auth success pulls default theme and toggles as-needed
4. auth'd app loads with user's default theme`theme`auth success pulls default theme and toggles as-needed
5. auth'd app loads with user's default theme`default dark used
6. auth success pulls default theme and toggles as-needed
7. auth'd app loads with user's default theme

does this addition suggest that the theme context is rebuilt whenever auth changes? wouldn't that be redundant with toggling theme in login success call? or perhaps at least the remove the login success theme toggle in favor of this?

**Review by anthonybernardi — APPROVED** [review]
nice
